src/plone/api/relation.py

```diff
@@ -0,0 +1,230 @@
+        raise InvalidParameterError('{} is no dexterity content'.format(source))
+
+    if target is not None and not IDexterityContent.providedBy(target):
+        raise InvalidParameterError('{} is no dexterity content'.format(target))
+
+    if relationship is not None and not isinstance(relationship, six.string_types):
+        raise InvalidParameterError('{} is no string'.format(relationship))
+
+    query = {}
+    relation_catalog = getUtility(ICatalog)
+    intids = getUtility(IIntIds)
+    if source is not None:
+        query['from_id'] = intids.getId(source)
+    if target is not None:
+        query['to_id'] = intids.getId(target)
+    if relationship is not None:
+        query['from_attribute'] = relationship
+    # If the query is empty, we could do relation_catalog.clear().
+    for rel in relation_catalog.findRelations(query):
+        relation_catalog.unindex(rel)
+
+
+@at_least_one_of('source', 'target', 'relationship')
+def get(source=None, target=None, relationship=None,
+        unrestricted=False, as_dict=False):
+    """Get specific relations given a source/target/relationship
+
+    Copied and modified from collective.relationhelpers get_relations.
+    """
+    if source is not None and not IDexterityContent.providedBy(source):
+        raise InvalidParameterError('{} is no dexterity content'.format(source))
+
+    if target is not None and not IDexterityContent.providedBy(target):
+        raise InvalidParameterError('{} is no dexterity content'.format(target))
+
+    if relationship is not None and not isinstance(relationship, six.string_types):
+        raise InvalidParameterError('{} is no string'.format(relationship))
+
+    intids = getUtility(IIntIds)
+    relation_catalog = getUtility(ICatalog)
+    query = {}
+    results = []
+
+    if as_dict:
+        results = defaultdict(list)
+
+    if not relation_catalog:
+        return results
+
+    if not unrestricted:
+        checkPermission = getSecurityManager().checkPermission
+
+    if source is not None:
+        query['from_id'] = intids.getId(source)
+    if target is not None:
+        query['to_id'] = intids.getId(target)
+    if relationship is not None:
+        query['from_attribute'] = relationship
+
+    for relation in relation_catalog.findRelations(query):
+        if relation.isBroken():
+            continue
+
+        if not unrestricted:
+            source_obj = relation.from_object
+            target_obj = relation.to_object
+
+            if checkPermission('View', source_obj) and checkPermission('View',
+                    target_obj):
+                if as_dict:
+                    results[relation.from_attribute].append(relation)
+                else:
+                    results.append(relation)
+            else:
+                continue
+        else:
+            if as_dict:
+                results[relation.from_attribute].append(relation)
+            else:
+                results.append(relation)
+    return results
```

Comment:
DRY: I think I saw something very similar in the controlpanel code.

Reply:
According to Philip the similar code in the controlpanel is from an earlier version, and the code is no longer there.